### ISSUE TITLE
Confusing handling of return value of start_unit(), since it returns EXIT_* which is >= 0

### ISSUE THREAD

**Issue by filbranden** [systemd/systemd#9704]
I found some confusing handling of return value of start_unit() while working on #9675.

In particular, [here](https://github.com/systemd/systemd/blob/8b3169b9a171b7b5255fa7c1b6f84cd324cf6397/src/systemctl/systemctl.c#L3121):

```c
                q = start_unit_one(bus, method, *name, mode, &error, w, arg_wait ? &wait_context : NULL);
                if (r >= 0 && q < 0)
                        r = translate_bus_error_to_exit_status(q, &error);
```

And [here](https://github.com/systemd/systemd/blob/8b3169b9a171b7b5255fa7c1b6f84cd324cf6397/src/systemctl/systemctl.c#L3152):

```c
        if (r >= 0 && arg_wait) {
```

The problem with that code is that, before it, `r` is only set [here](https://github.com/systemd/systemd/blob/8b3169b9a171b7b5255fa7c1b6f84cd324cf6397/src/systemctl/systemctl.c#L3122):

```c
                        r = translate_bus_error_to_exit_status(q, &error);
```

Which will set `r` to one of the `EXIT_*` values, which means 0 for success and >0 for one of the error exists... So the check `r >= 0` is really meaningless and I guess what should be tested there is `r == 0` for "no error was set, yet."

The first occurrence was introduced in commit e3e0314b56012f7febc279d268f2cadc1fcc0f25, which has in fact changed it from `if (r == 0 && q < 0)` to `if (r >= 0 && q < 0)`, which makes me wonder if this was somehow intentional (?!)

The second one was introduced in 93a0884126146361ca078ec627da2cf766205a1c (which implements `--wait` option), but I wonder if that one was due to a copy & paste of the code above it.

(Another possibility is that `translate_bus_error_to_exit_status()` may return <0 if the error is unknown... Is that really what we're checking here?)

Anyways, just wanted to raise an issue, to double check that this is indeed all intended or not. cc @keszybz who originally authored the code in e3e0314b56012f7febc279d268f2cadc1fcc0f25 and also the `if (r == 0 && q < 0)` which preceded it (commit b65205466248531d52f9d6dba16195864b023d4a).

Cheers,
Filipe

**Comment by yuwata** [comment]
Yeah, I was also confused when I tried to fix a corresponding LGTM warning...

**Comment by poettering** [comment]
Fix waiting in #10391.
